README_python.md

```diff
@@ -170,68 +201,294 @@ essentially zero cost, with storage being shared. Thus, we can usually pass it
 directly to existing Python APIs (like `imsave()`) that expect 'image-like'
 objects without any explicit conversion necessary.
 
-## Using Halide Generators from Python
+## Halide Generators In Python
+
+In Halide, a "Generator" is a unit of encapsulation for Halide code. It is
+self-contained piece of code that can:
+-  Produce a chunk of Halide IR (in the form of an `hl.Pipeline`) that is appropriate for compilation (via either JIT or
+AOT)
+-  Expose itself to the build system in a discoverable way
+-  Fully describe
+itself for the build system with metadata for (at least) the type and number of
+inputs and outputs expected
+-  Allow for build-time customization of
+coder-specified parameters in a way that doesn't require editing of source code
+
+Originally, Halide only supported writing Generators in C++. In this document,
+we'll use the term "C++ Generator" to mean "Generator written in C++ using the
+classic API", the term "Python Generator" to mean "Generator written in Halide's
+Python bindings", and just plain "Generator" when the discussion is relatively
+neutral with respect to the implementation language/API.
+
+### Writing a Generator in Python
+
+A Python Generator is a class that: - has the `@hl.generator` decorator applied
+to it - declares zero or more member fields that are initialized with values of
+`hl.InputBuffer` or `hl.InputScalar`, which specify the expected input(s) of the
+resulting `Pipeline`. - declares one or more member fields that are initialized
+with values of `hl.OutputBuffer` or `hl.OutputScalar`, which specify the
+expected output(s) of the resulting `Pipeline`. - declares zero or more member
+fields that are initialized with values of `hl.GeneratorParam`, which can be
+used to pass arbitrary information from the build system to the Generator. A
+GeneratorParam can carry a value of type `bool`, `int`, `float`, `str`, or
+`hl.Type`. - declares a `generate()` method that fill in the Halide IR needed to
+define all of the Outputs - optionally declares a `configure()` method to
+dynamically add Inputs or Outputs to the pipeline, based on (e.g.) the values of
+`GeneratorParam` values or other external inputs
+
+Let's look at a fairly simple example:
```

Comment:
The main thing missing here is a standard guideline for documentation. I'm not sure if this should be a longer comment above the class that describes all the parameters, comments on the parameter lines themselves, or a docstring in the definition of the class, but I am certain one should never write a generator without documentation on what it does and what each of its parameters, inputs, and outputs are.

Reply:
Good point! I'll prep that as a PR-onto-this-PR, as it may need separate comment.